util/pkg/vfs/s3context.go

```diff
@@ -106,3 +100,49 @@ func (s *S3Context) getRegionForBucket(bucket string) (string, error) {
 
 	return region, nil
 }
+
+/*
+Amazon's S3 API provides the GetBucketLocation call to determine the region in which a bucket is located.
+This call can however only be used globally by the owner of the bucket, as mentioned on the documentation page.
+
+For S3 buckets that are shared across multiple AWS accounts using bucket policies the call will only work if it is sent
+to the correct region in the first place.
+
+This method will attempt to "bruteforce" the bucket location by sending a request to every available region and picking
+out the first result.
+
+See also: https://docs.aws.amazon.com/goto/WebAPI/s3-2006-03-01/GetBucketLocationRequest
+*/
+func bruteforceBucketLocation(region *string, bucket *string) (*s3.GetBucketLocationOutput, error) {
+	session, _ := session.NewSession(&aws.Config{Region: region})
+	regions, err := ec2.New(session).DescribeRegions(nil)
+
+	if err != nil {
+		return nil, fmt.Errorf("Unable to list AWS regions: %v", err)
+	}
+
+	request := &s3.GetBucketLocationInput{
+		Bucket: bucket,
+	}
+
+	glog.V(2).Infof("Querying S3 for bucket location for %s", *bucket)
+
+	out := make(chan *s3.GetBucketLocationOutput)
+	for _, region := range regions.Regions {
+		go func(regionName string) {
+			s3Client := s3.New(session, &aws.Config{Region: aws.String(regionName)})
+			result, bucketError := s3Client.GetBucketLocation(request)
+
+			if bucketError == nil {
+				out <- result
+			}
+		}(*region.RegionName)
+	}
+
+	select {
+	case bucketLocation := <-out:
+		return bucketLocation, nil
```

Comment:
I don't see a reason to, *if* we get a response and no error the response will contain the correct location (also in cases of getting the "proper" response as a bucket owner, i.e. globally).